# Lab | Hypothesis Testing

**Objective**

Welcome to the Hypothesis Testing Lab, where we embark on an enlightening journey through the realm of statistical decision-making! In this laboratory, we delve into various scenarios, applying the powerful tools of hypothesis testing to scrutinize and interpret data.

From testing the mean of a single sample (One Sample T-Test), to investigating differences between independent groups (Two Sample T-Test), and exploring relationships within dependent samples (Paired Sample T-Test), our exploration knows no bounds. Furthermore, we'll venture into the realm of Analysis of Variance (ANOVA), unraveling the complexities of comparing means across multiple groups.

So, grab your statistical tools, prepare your hypotheses, and let's embark on this fascinating journey of exploration and discovery in the world of hypothesis testing!

**Challenge 1**

In this challenge, we will be working with pokemon data. The data can be found here:

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv

In [1]:
#libraries
import pandas as pd
import scipy.stats as st
import numpy as np



In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv")
df

,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,Mega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
...,...,...,...,...,...,...,...,...,...,...,...
795,Diancie,Rock,Fairy,50,100,150,100,150,50,6,True
796,Mega Diancie,Rock,Fairy,50,160,110,160,110,110,6,True
797,Hoopa Confined,Psychic,Ghost,80,110,60,150,130,70,6,True
798,Hoopa Unbound,Psychic,Dark,80,160,60,170,130,80,6,True


- We posit that Pokemons of type Dragon have, on average, more HP stats than Grass. Choose the propper test and, with 5% significance, comment your findings.

In [3]:
df.columns.tolist()

['Name',
 'Type 1',
 'Type 2',
 'HP',
 'Attack',
 'Defense',
 'Sp. Atk',
 'Sp. Def',
 'Speed',
 'Generation',
 'Legendary']

In [4]:
#code here
dragon_hp = df[df["Type 1"] == "Dragon"]["HP"].dropna()
grass_hp = df[df["Type 1"] == "Grass"]["HP"].dropna()

In [5]:
print("Number of Dragon Pokémon:", len(dragon_hp))
print("Number of Grass Pokémon:", len(grass_hp))

print("Dragon mean HP:", dragon_hp.mean())
print("Grass mean HP:", grass_hp.mean())

Number of Dragon Pokémon: 32
Number of Grass Pokémon: 70
Dragon mean HP: 83.3125
Grass mean HP: 67.27142857142857


In [6]:
from scipy.stats import ttest_ind

alpha = 0.05

t_statistic, p_value = ttest_ind(
    dragon_hp,
    grass_hp,
    equal_var=False,
    alternative="greater"
)

print("T-statistic:", t_statistic)
print("P-value:", p_value)

T-statistic: 3.3349632905124063
P-value: 0.0007993609745420598


In [7]:
if p_value < alpha:
    print("Reject the null hypothesis.")
    print("Dragon Pokémon have significantly higher average HP than Grass Pokémon.")
else:
    print("Fail to reject the null hypothesis.")
    print("There is not enough evidence that Dragon Pokémon have higher average HP than Grass Pokémon.")

Reject the null hypothesis.
Dragon Pokémon have significantly higher average HP than Grass Pokémon.


- We posit that Legendary Pokemons have different stats (HP, Attack, Defense, Sp.Atk, Sp.Def, Speed) when comparing with Non-Legendary. Choose the propper test and, with 5% significance, comment your findings.


In [19]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv")
df

,Name,Type 1,Type 2,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation,Legendary
0,Bulbasaur,Grass,Poison,45,49,49,65,65,45,1,False
1,Ivysaur,Grass,Poison,60,62,63,80,80,60,1,False
2,Venusaur,Grass,Poison,80,82,83,100,100,80,1,False
3,Mega Venusaur,Grass,Poison,80,100,123,122,120,80,1,False
4,Charmander,Fire,NaN,39,52,43,60,50,65,1,False
...,...,...,...,...,...,...,...,...,...,...,...
795,Diancie,Rock,Fairy,50,100,150,100,150,50,6,True
796,Mega Diancie,Rock,Fairy,50,160,110,160,110,110,6,True
797,Hoopa Confined,Psychic,Ghost,80,110,60,150,130,70,6,True
798,Hoopa Unbound,Psychic,Dark,80,160,60,170,130,80,6,True


In [20]:
#code here
stats_columns = [
    "HP",
    "Attack",
    "Defense",
    "Sp. Atk",
    "Sp. Def",
    "Speed", 
]

In [21]:
print(df.columns.tolist())

print([
    column
    for column in df.columns
    if "legend" in column.lower()
])

['Name', 'Type 1', 'Type 2', 'HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary']
['Legendary']


In [22]:
from scipy.stats import ttest_ind

stats_columns = [
    "HP",
    "Attack",
    "Defense",
    "Sp. Atk",
    "Sp. Def",
    "Speed"
]

for column in stats_columns:
    legendary = df.loc[df["Legendary"] == True, column].dropna()
    non_legendary = df.loc[df["Legendary"] == False, column].dropna()

    t_stat, p_value = ttest_ind(
        legendary,
        non_legendary,
        equal_var=False
    )

    print(f"{column}: t = {t_stat:.3f}, p = {p_value:.4f}")

HP: t = 8.981, p = 0.0000
Attack: t = 10.438, p = 0.0000
Defense: t = 7.637, p = 0.0000
Sp. Atk: t = 13.417, p = 0.0000
Sp. Def: t = 10.016, p = 0.0000
Speed: t = 11.475, p = 0.0000


Independent two-sample Welch’s t-tests were conducted to compare the average base statistics of Legendary and non-Legendary Pokémon.

For all six statistics, the p-values were below the significance level of 0.05. Although the output displays the p-values as 0.0000, this does not mean the values are exactly zero; it means they are extremely small and have been rounded to four decimal places.

Therefore, the null hypothesis is rejected for every statistic. There is strong statistical evidence that Legendary and non-Legendary Pokémon differ significantly in terms of HP, Attack, Defense, Special Attack, Special Defense, and Speed.

All t-statistics were positive because the Legendary Pokémon group was entered first in the test. This indicates that Legendary Pokémon had higher average values than non-Legendary Pokémon for all six statistics.

The largest difference relative to the variation in the data was observed for Special Attack, which had the highest t-statistic:

HP: (t = 8.981,\ p < 0.0001)
Attack: (t = 10.438,\ p < 0.0001)
Defense: (t = 7.637,\ p < 0.0001)
Special Attack: (t = 13.417,\ p < 0.0001)
Special Defense: (t = 10.016,\ p < 0.0001)
Speed: (t = 11.475,\ p < 0.0001)

Overall, the analysis suggests that Legendary Pokémon are statistically stronger than non-Legendary Pokémon across all six base-stat categories. However, statistical significance alone does not indicate how large the differences are. Comparing the group means or calculating an effect size would help determine the practical magnitude of these differences.

**Challenge 2**

In this challenge, we will be working with california-housing data. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv

In [26]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv")
df.head()

,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
0,-114.31,34.19,15.0,5612.0,1283.0,1015.0,472.0,1.4936,66900.0
1,-114.47,34.40,19.0,7650.0,1901.0,1129.0,463.0,1.8200,80100.0
2,-114.56,33.69,17.0,720.0,174.0,333.0,117.0,1.6509,85700.0
3,-114.57,33.64,14.0,1501.0,337.0,515.0,226.0,3.1917,73400.0
4,-114.57,33.57,20.0,1454.0,326.0,624.0,262.0,1.9250,65500.0


**We posit that houses close to either a school or a hospital are more expensive.**

- School coordinates (-118, 34)
- Hospital coordinates (-122, 37)

We consider a house (neighborhood) to be close to a school or hospital if the distance is lower than 0.50.

Hint:
- Write a function to calculate euclidean distance from each house (neighborhood) to the school and to the hospital.
- Divide your dataset into houses close and far from either a hospital or school.
- Choose the propper test and, with 5% significance, comment your findings.
 

# create a Disance Function

In [27]:
def distance_function(
    house_longitude,
    house_latitude,
    location_longitude,
    location_latitude
):
    longitude_difference = house_longitude - location_longitude
    latitude_difference = house_latitude - location_latitude

    distance = (
        longitude_difference**2
        + latitude_difference**2
    )**0.5

    return distance

In [28]:
distance_function(-118, 34, -118, 34)

0.0

In [30]:
school_longitude = -118
school_latitude = 34

hospital_longitude = -122
hospital_latitude = 37

In [31]:
df["distance_to_school"] = df.apply(
    lambda row: distance_function(
        row["longitude"],
        row["latitude"],
        school_longitude,
        school_latitude
    ),
    axis=1
)

In [40]:
df["distance_to_hospital"] = df.apply(
    lambda row: distance_function(
        row["longitude"],
        row["latitude"],
        hospital_longitude,
        hospital_latitude
    ),
    axis=1
)

In [41]:
df[
    [
        "longitude",
        "latitude",
        "distance_to_school",
        "distance_to_hospital"
    ]
].head()

,longitude,latitude,distance_to_school,distance_to_hospital
0,-114.31,34.19,3.694888,8.187319
1,-114.47,34.40,3.552591,7.966235
2,-114.56,33.69,3.453940,8.143077
3,-114.57,33.64,3.448840,8.154416
4,-114.57,33.57,3.456848,8.183508


In [42]:
df["close_to_facility"] = (
    (df["distance_to_school"] < 0.50)
    |
    (df["distance_to_hospital"] < 0.50)
)

In [43]:
df["close_to_facility"].value_counts()

close_to_facility
False    10171
True      6829
Name: count, dtype: int64

In [44]:
df.columns.tolist()

['longitude',
 'latitude',
 'housing_median_age',
 'total_rooms',
 'total_bedrooms',
 'population',
 'households',
 'median_income',
 'median_house_value',
 'distance_to_school',
 'distance_to_hospital',
 'close_to_facility']

In [45]:
close_prices = df.loc[
    df["close_to_facility"] == True,
    "median_house_value"
].dropna()

far_prices = df.loc[
    df["close_to_facility"] == False,
    "median_house_value"
].dropna()

In [46]:
print("Number of close houses:", len(close_prices))
print("Number of far houses:", len(far_prices))

print("Average price of close houses:", close_prices.mean())
print("Average price of far houses:", far_prices.mean())

Number of close houses: 6829
Number of far houses: 10171
Average price of close houses: 246951.98213501245
Average price of far houses: 180678.44105790975


In [47]:
from scipy.stats import ttest_ind

t_stat, p_value = ttest_ind(
    close_prices,
    far_prices,
    equal_var=False,
    alternative="greater"
)

print(f"t-statistic: {t_stat:.3f}")
print(f"p-value: {p_value:.4f}")

t-statistic: 37.992
p-value: 0.0000


In [48]:
alpha = 0.05

if p_value < alpha:
    print(
        "Reject the null hypothesis. There is statistically "
        "significant evidence that houses close to a school "
        "or hospital are more expensive."
    )
else:
    print(
        "Fail to reject the null hypothesis. There is not "
        "enough evidence that houses close to a school or "
        "hospital are more expensive."
    )

Reject the null hypothesis. There is statistically significant evidence that houses close to a school or hospital are more expensive.
